service/history/callback_queue_factory.go

```diff
@@ -0,0 +1,245 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	"net/http"
+	"sync"
+
+	"go.uber.org/fx"
+
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/quotas"
+	ctasks "go.temporal.io/server/common/tasks"
+	"go.temporal.io/server/service/history/queues"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tasks"
+	wcache "go.temporal.io/server/service/history/workflow/cache"
+)
+
+const callbackQueuePersistenceMaxRPSRatio = 0.3
+
+type callbackQueueFactoryParams struct {
+	fx.In
+
+	QueueFactoryBaseParams
+}
+
+type groupPool[K comparable, T any] struct {
+	mu        sync.RWMutex
+	pool      map[K]T
+	construct func(K) T
+}
+
+func newGroupPool[K comparable, T any](construct func(K) T) *groupPool[K, T] {
+	return &groupPool[K, T]{
+		construct: construct,
+		pool:      make(map[K]T, 0),
+	}
+}
+
+func (p *groupPool[K, T]) getOrCreate(key K) T {
+	p.mu.RLock()
+	value, ok := p.pool[key]
+	p.mu.RUnlock()
+	if !ok {
+		p.mu.Lock()
+		if value, ok = p.pool[key]; !ok {
+			value = p.construct(key)
+		}
+		p.mu.Unlock()
+	}
+
+	return value
+}
+
+// TODO: get actual limits from dynamic config.
+type groupLimiter struct {
+	key queues.NamespaceIDAndDestination
+}
+
+func (groupLimiter) BufferSize() int {
+	return 100
+}
+
+func (groupLimiter) Concurrency() int {
+	return 100
+}
+
+var _ ctasks.DynamicWorkerPoolLimiter = groupLimiter{}
+
+type callbackQueueFactory struct {
+	callbackQueueFactoryParams
+	hostReaderRateLimiter quotas.RequestRateLimiter
+	// Shared client pool for all shards in the host.
+	completionCallerPool *groupPool[queues.NamespaceIDAndDestination, HTTPCaller]
+	// Shared rate limiter pool for all shards in the host.
+	rateLimiterPool *groupPool[queues.NamespaceIDAndDestination, quotas.RateLimiter]
+	// Shared scheduler across all shards in the host.
+	hostScheduler queues.Scheduler
+}
+
+func NewCallbackQueueFactory(params callbackQueueFactoryParams) QueueFactory {
+	rateLimiterPool := newGroupPool(func(queues.NamespaceIDAndDestination) quotas.RateLimiter {
+		// TODO: get this value from dynamic config.
```

Comment:
Just curious, will we need to add support for getting dynamic config values by arbitrary key (in this case destination)? AFAIK we only have namespace and task queue filtered properties atm.

Reply:
Still not sure how we'll store this config. We might need this capability.